src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -723,27 +722,21 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             !alloc = bloomFilterAllocForLevel conf ln
             !runPaths = Paths.runPath root (uniqueToRunNumber n)
         traceWith tr $ AtLevel ln $
-          TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergelast
-        let numInputRuns = NumRuns $ V.length rs
-        let numInputEntries = V.foldMap' Run.size rs
+          TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergeLevel
+        -- TODO: creating the merge should happen in MR.new
+        mergeMaybe <- allocateMaybeTemp reg
+          (Merge.new hfs hbio caching alloc mergeLevel resolve runPaths rs)
+          Merge.abort
+        mr <- case mergeMaybe of
+          Nothing -> error "newMerge: merges can not be empty"
+          Just m  -> allocateTemp reg (MR.new mergePolicy rs m) releaseRef
         case confMergeSchedule of
+          Incremental -> pure ()
           OneShot -> do
-            r <- allocateTemp reg
-                  (mergeRuns resolve hfs hbio caching alloc runPaths mergelast rs)
-                  releaseRef
-            traceWith tr $ AtLevel ln $
-              TraceCompletedMerge (Run.size r)
-                                  (Run.runFsPathsNumber r)
-            V.mapM_ (freeTemp reg . releaseRef) rs
-            Merging <$!> MR.unsafeNew mergePolicy numInputRuns numInputEntries MR.MergeKnownCompleted (MR.CompletedMerge r)
-
-          Incremental -> do
-            mergeMaybe <- allocateMaybeTemp reg
-              (Merge.new hfs hbio caching alloc mergelast resolve runPaths rs)
-              Merge.abort
-            case mergeMaybe of
-              Nothing -> error "newMerge: merges can not be empty"
-              Just m  -> Merging <$!> MR.new mergePolicy rs m
+            let !required = MR.Credits (unNumEntries (V.foldMap' Run.size rs))
+            let !thresh = creditThresholdForLevel conf ln
+            MR.supplyCredits required thresh mr
```

Comment:
Joris makes a fair point. And it's not something we'd notice easily since it'd in practice do 99.9% of the work in one shot. And maybe that's ok? Or do we need to force spending of the credits and assert it's complete?

Reply:
Yes, I remember contemplating that at some point, and yes, I do think it's the right thing to really complete the merge immediately instead of only 99.9%.

I think it works correctly here. We had a related discussion when `supplyMergeCredits` was added, which lead to adding the "We can finish the merge immediately" branch. I think that should take care of it.

But since it's not really obvious, I should at least add an assert to be sure.

Reply:
It indeed passes the assertion, but fails if I supply `required - 1` credits. Interestingly, I get a RefCount assertion failure, which is probably because the failing `isComplete` assertion leads to rolling back the temp registry, triggering the bug we discussed in the other comment. :smile:

Reply:
Instead of the slightly awkward assertion on the MergingRun's internal state, I decided to make `expectCompleted` not release the merging run, so we can use it here to 1. make sure the merge is completed and 2. access the resulting run's size for tracing.

https://github.com/IntersectMBO/lsm-tree/pull/498/commits/90cac301f287c6b39ffdf369bb205eec9cc1114c